src/mbgl/shaders/gl/shader_program_gl.cpp

```diff
@@ -0,0 +1,148 @@
+    }
+}
+
+ShaderProgramGL::ShaderProgramGL(UniqueProgram&& glProgram_)
+    : ShaderProgramBase(),
+      glProgram(std::move(glProgram_)) {
+}
+
+ShaderProgramGL::ShaderProgramGL(UniqueProgram&& program,
+                                 VertexAttributeArrayGL&& uniforms_,
+                                 VertexAttributeArrayGL&& attributes_)
+    : ShaderProgramBase(),
+      glProgram(std::move(program)),
+      uniforms(std::move(uniforms_)),
+      vertexAttributes(std::move(attributes_)) {
+}
+
+ShaderProgramGL::ShaderProgramGL(ShaderProgramGL&& other)
+    : ShaderProgramBase(std::forward<ShaderProgramBase&&>(other)),
+      glProgram(std::move(other.glProgram)),
+      uniforms(std::move(other.uniforms)),
+      vertexAttributes(std::move(other.vertexAttributes)) {
+}
+
+using namespace platform;
+
+static void addAttr(VertexAttributeArrayGL& attrs, const char* name, GLint index,
+                    GLsizei length, GLint count, GLenum glType) {
+    const auto elementType = mapType(glType);
+    const auto elementCount = mapCount(glType); // number of `elementType`, hopefully temporary
+    if (elementType != gfx::AttributeDataType::Invalid && length > 0) {
+        if (auto newAttr = attrs.add(name, index, elementType, elementCount, count)) {
+            static_cast<VertexAttributeGL*>(newAttr)->setGLType(glType);
+        }
+    }
+}
+
+std::shared_ptr<ShaderProgramGL> ShaderProgramGL::create(
+        Context& context,
+        std::string_view /*name*/,
+        std::string_view vertexSource,
+        std::string_view fragmentSource) noexcept(false) {
+    
+    const auto firstAttrib = "pos";
+
+    // throws on compile error
+    auto vertProg = context.createShader(ShaderType::Vertex, std::initializer_list<const char*>{vertexSource.data()});
+    auto fragProg = context.createShader(ShaderType::Fragment, std::initializer_list<const char*>{fragmentSource.data()});
+    auto program = context.createProgram(vertProg, fragProg, firstAttrib);
+
+    // GLES3.1
+    //GLint numAttribs;
+    //glGetProgramInterfaceiv(program, GL_PROGRAM_INPUT, GL_ACTIVE_RESOURCES, &numAttribs);
+
+    VertexAttributeArrayGL uniforms;
+
+    GLint count = 0;
+    GLint maxLength = 0;
+    MBGL_CHECK_ERROR(glGetProgramiv(program, GL_ACTIVE_UNIFORMS, &count));
+    MBGL_CHECK_ERROR(glGetProgramiv(program, GL_ACTIVE_UNIFORM_MAX_LENGTH, &maxLength));
+
+    auto name = std::make_unique<GLchar[]>(maxLength);
+    for (GLint index = 0; index < count; ++index) {
+        GLsizei length = 0;
+        GLint size = 0;
+        GLenum glType = 0;
+        MBGL_CHECK_ERROR(glGetActiveUniform(program, index, maxLength, &length, &size, &glType, name.get()));
+        addAttr(uniforms, name.get(), index, length, size, glType);
+    }
+
+    VertexAttributeArrayGL attrs;
+
+    count = 0;
+    MBGL_CHECK_ERROR(glGetProgramiv(program, GL_ACTIVE_ATTRIBUTES, &count));
+    MBGL_CHECK_ERROR(glGetProgramiv(program, GL_ACTIVE_ATTRIBUTE_MAX_LENGTH, &maxLength));
```

Comment:
Reallocate `name` here, otherwise when GL_ACTIVE_ATTRIBUTE_MAX_LENGTH returns a size greater than GL_ACTIVE_UNIFORM_MAX_LENGTH you're corrupting the heap.